remotehttp.go

```diff
@@ -95,94 +97,105 @@ func (r *RemoteHTTPBase) String() string {
-			"method": "GET",
-			"url":    u.String(),
+		resp *http.Response
+		log  = Log.WithFields(logrus.Fields{
+			"method":  method,
+			"url":     u.String(),
+			"attempt": attempt,
 		})
 	)
-retry:
-	attempt++
-	req, err := http.NewRequest("GET", u.String(), nil)
+
+	req, err := http.NewRequest(method, u.String(), getReader())
 	if err != nil {
-		return nil, err
+		log.Debug("unable to create new request")
+		return 0, nil, err
 	}
 	if r.opt.HTTPAuth != "" {
 		req.Header.Set("Authorization", r.opt.HTTPAuth)
 	}
-	log.Debug("request sent")
+
+	log.Debug("sending request")
 	resp, err = r.client.Do(req)
 	if err != nil {
-		if attempt >= r.opt.ErrorRetry {
-			log.WithField("attempt", attempt).WithError(err).Error("failed, giving up")
-			return nil, errors.Wrap(err, u.String())
-		}
-		log.WithField("attempt", attempt).WithError(err).Error("failed, retrying")
-		goto retry
+		log.WithError(err).Error("error while sending request")
+		return 0, nil, errors.Wrap(err, u.String())
 	}
-	log.WithField("status", resp.StatusCode).Debug("response received")
+
 	defer resp.Body.Close()
-	switch resp.StatusCode {
-	case 200: // expected
-	case 404:
-		return nil, NoSuchObject{name}
-	default:
-		return nil, fmt.Errorf("unexpected status code %d from %s", resp.StatusCode, name)
-	}
+
 	b, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		if attempt >= r.opt.ErrorRetry {
-			log.WithField("attempt", attempt).WithError(err).Error("failed, giving up")
-			return nil, errors.Wrap(err, u.String())
-		}
-		log.WithField("attempt", attempt).WithError(err).Error("failed, retrying")
-		goto retry
+		log.WithError(err).Error("error while reading response")
+		return 0, nil, errors.Wrap(err, u.String())
 	}
-	return b, nil
+
+	log.WithField("statusCode", resp.StatusCode).Debug("response received")
+	return resp.StatusCode, b, nil
 }
 
-// StoreObject stores an object to the store.
-func (r *RemoteHTTPBase) StoreObject(name string, rdr io.Reader) error {
+// Send a single HTTP request, retrying if a retryable error has occurred.
+func (r *RemoteHTTPBase) IssueRetryableHttpRequest(method string, u *url.URL, getReader GetReaderForRequestBody) (int, []byte, error) {
 
-	u, _ := r.location.Parse(name)
 	var (
-		resp    *http.Response
-		err     error
 		attempt int
 		log     = Log.WithFields(logrus.Fields{
-			"method": "PUT",
+			"method": method,
 			"url":    u.String(),
 		})
 	)
+
 retry:
 	attempt++
-	req, err := http.NewRequest("PUT", u.String(), rdr)
+	statusCode, responseBody, err := r.IssueHttpRequest(method, u, getReader, attempt)
+
+	if (err != nil) || (statusCode >= 500 && statusCode < 600) {
+		if attempt >= r.opt.ErrorRetry {
+			log.WithField("attempt", attempt).Debug("failed, giving up")
+			return 0, nil, err
+		} else {
+			log.WithField("attempt", attempt).WithField("delay", attempt).Debug("waiting, then retrying")
+			baseInterval := r.opt.ErrorRetryBaseInterval
+			if baseInterval == 0 {
+				baseInterval = time.Duration(0)
+			}
```

Comment:
no need for the condition, but won't hurt either. Will remove it at some point, depending on what solution we'll go with for the units of base-interval